components/nonmem/src/metadata.rs

```
@@ -0,0 +1,235 @@
+            description,
+            tags: Vec::new(),
+        })
+    }
+
+    pub fn load(path: impl AsRef<Path>) -> Result<Self> {
+        let content = fs::read_to_string(path)?;
+        Ok(serde_json::from_str(&content)?)
+    }
+
+    pub fn save(&self, model_name: &str, folder: impl AsRef<Path>) -> Result<()> {
+        if self.description.trim().is_empty() {
+            bail!(
+                "No description was found in the metadata file. Please 'set' or 'append' a description."
+            )
+        }
+
+        let metadata_path = folder
+            .as_ref()
+            .join(format!("{model_name}{METADATA_FILENAME_SUFFIX}"));
+        write_json_to_file(self, metadata_path)?;
+        Ok(())
+    }
+
+    pub fn update(
+        mut self,
+        description: Option<String>,
+        mut tags: Vec<String>,
+        mut based_on: Vec<String>,
+        overwrite: bool,
+    ) -> Self {
+        if overwrite {
+            // Overwrite mode: replace fields that are provided
+            if let Some(d) = description
+                && !d.trim().is_empty()
+            {
+                self.description = d;
+            }
+            if !tags.is_empty() {
+                self.tags = tags;
+            }
+            if !based_on.is_empty() {
+                self.based_on = based_on;
+            }
+        } else {
+            // Append mode: merge with existing
+            if !tags.is_empty() {
+                self.tags.append(&mut tags);
+            }
+
+            let tags_set = self.tags.into_iter().collect::<HashSet<String>>();
+            self.tags = tags_set.into_iter().collect();
+
+            if !based_on.is_empty() {
+                self.based_on.append(&mut based_on);
+            }
+
+            let based_set = self.based_on.into_iter().collect::<HashSet<String>>();
+            self.based_on = based_set.into_iter().collect();
+
+            if let Some(d) = description
+                && !self.description.contains(&d)
```

Any particular reasons for those checks here and below and not just straight appending (maybe with a space if it doesn't end with one)?  That will silently ignore or change the description